include/boost/proto/detail/expr.hpp

```diff
@@ -138,6 +138,24 @@
         BOOST_PP_REPEAT(ARG_COUNT, BOOST_PROTO_CHILD, ~)
         BOOST_PP_REPEAT_FROM_TO(ARG_COUNT, BOOST_PROTO_MAX_ARITY, BOOST_PROTO_VOID, ~)
 
+        #define BOOST_PROTO_HASH #
+
+        #if defined(__WAVE__)
+        BOOST_PROTO_HASH ifndef BOOST_NO_CXX11_DEFAULTED_FUNCTIONS
+        #endif
+        #if !defined(BOOST_NO_CXX11_DEFAULTED_FUNCTIONS) || defined(__WAVE__)
+            /// Define default constructor and copy constructor as explicitly defaulted
+            /// if the compiler supports it. This allows to silence gcc's deprecated-copy
+            /// warnings.
+            BOOST_FORCEINLINE expr() = default;
+            BOOST_FORCEINLINE expr(const expr&) = default;
```

Comment:
This suppresses the generation of the move constructor.

Reply:
> Requesting changes...

Hi,

I have implemented these changes.

Cheers,
Romain

Reply:
I'm also concerned that it has the effect of suppressing the generation of the built-in `operator=`. That is potentially behavior-changing.

On deeper reflection, I'm not comfortable with this change at all. Sorry.

Reply:
Well, the gcc warning is all about foreseeing an upcoming change mandated by C++ standard, where defining the copy operator explictely will eventually prevent the implicit copy constructor to be issued. So kind of the same things you are worried about here. So if not fixed now, the problem will eventually come.

If these changes are not ok, would it be acceptable to silence the gcc warnings for now ?

Reply:
The other issue with this change is that it changes `expr` from being an aggregate to a non-aggregate. I suspect that interferes with aggregate initialization. But that would break everything. I'm confused how this works.

Reply:
Hi @ericniebler,

When trying to migrate to C++20 and gcc 10, I am hitting again this issue, this time in expr_variadic.hpp. Since indeed my change would delete the aggregate constructor (some subtleties I actually didn't know), I guess the only short term option would be to simply silence the gcc warnings. I am not sure if that works (I don't know if gcc pragmas are taken at the location the class is defined or at the location the class is used). If it works, would you be ok for sur a pull request.

Anyway, C++ is deprecating the usage of implicit copy constructor when an explicit copy assigment operator exists, and such case are happening already. One day there is no doubt the standard will remove this behavior entirely, and that day a better solution than just silencing warnings will have to be found. Since you are somehow involved in the standardization process, I guess you are more knowledgeable than me to actually fix it when it comes ;)

Reply:
Apparently enclosing the whole file expr_variadic.hpp with pragma diagnostics ignoring `-Wdeprecated-copy` doesn't help, it has to be used at call site. In my case:

```
In file included from /data/mwrep/res/osp/Boost/20-0-0-0/include/boost/xpressive/xpressive_static.hpp:24,
                 from src/functional_tests/ApplicativeEvent.cpp:7: 
/data/mwrep/res/osp/Boost/20-0-0-0/include/boost/xpressive/regex_primitives.hpp: In static member function ‘static boost::xpressive::detail::basic_mark_tag boost::xpressive::mark_tag::make_tag(int)’: 
/data/mwrep/res/osp/Boost/20-0-0-0/include/boost/xpressive/regex_primitives.hpp:548:16: error: implicitly-declared ‘constexpr boost::proto::exprns_::expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::xpressive::detail::mark_placeholder>, 0>::expr(const boost::proto::exprns_::expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::xpressive::detail::mark_placeholder>, 0>&)’ is deprecated [-Werror=deprecated-copy]
  548 |         return mark;
      |                ^~~~
In file included from /data/mwrep/res/osp/Boost/20-0-0-0/include/boost/proto/detail/expr.hpp:4,
                 from /data/mwrep/res/osp/Boost/20-0-0-0/include/boost/proto/expr.hpp:137,
                 from /data/mwrep/res/osp/Boost/20-0-0-0/include/boost/proto/core.hpp:17,
                 from /data/mwrep/res/osp/Boost/20-0-0-0/include/boost/xpressive/regex_primitives.hpp:28,
                 from /data/mwrep/res/osp/Boost/20-0-0-0/include/boost/xpressive/xpressive_static.hpp:24,
                 from src/functional_tests/ApplicativeEvent.cpp:7: 
/data/mwrep/res/osp/Boost/20-0-0-0/include/boost/proto/detail/preprocessed/expr_variadic.hpp:99:9: note: because ‘boost::proto::exprns_::expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::xpressive::detail::mark_placeholder>, 0>’ has user-provided ‘const boost::proto::exprns_::expr<boost::proto::tagns_::tag::assign, boost::proto::argsns_::list2<boost::proto::exprns_::expr<Tag, boost::proto::argsns_::term<Arg>, 0>&, const boost::proto::exprns_::expr<Tag, boost::proto::argsns_::term<Arg>, 0>&>, 2> boost::proto::exprns_::expr<Tag, boost::proto::argsns_::term<Arg>, 0>::operator=(const boost::proto::exprns_::expr<Tag, boost::proto::argsns_::term<Arg>, 0>&) [with Tag = boost::proto::tagns_::tag::terminal; Arg0 = boost::xpressive::detail::mark_placeholder]’
   99 |         operator =(expr const &a)
      |         ^~~~~~~~
```

I am not sure if we want to start chasing down all users of proto and add these pragmas there.

What if in order to fix the aggregate issue we did explicitly define constructors taking the different member initializers as argument ? And what if the copy constructor would be protected by something like

```
proto::expr<proto::tag::assign, list2<expr &, expr const &>, 2> const
```

which you use already for the copy assignment operator ?

Cheers,
Romain

Reply:
@ericniebler Do you think it would make sense to ask standard people to find out some C++23 syntax so that we can express that we want to have the aggregate initialization constructor available and defaulted in a class, despite having defined other constructors (like the copy or the default one) ?

Reply:
> The other issue with this change is that it changes expr from being an aggregate to a non-aggregate. I suspect that interferes with aggregate initialization. But that would break everything. I'm confused how this works.

Every major compiler considers it an aggregate (https://godbolt.org/z/3K8jtv), even in C++17 mode, where it was made non-aggregate (https://eel.is/c++draft/diff.cpp17.dcl.dcl#3).

Reply:
Ah, the change is for C++20, and I was thinking it is retroactive while it seems to be not.

Reply:
I filed a bug about this case https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94492

Reply:
On our side, while this patch worked for us while we were using C++17 (maybe it was wrong, but in the use cases we had it did not show issue), however when we moved to C++20 indeed the fact that the "aggregate constructor" wasn't allowed anymore, it indeed broke the rest of the boost libraries using it.

We internally decided to stop patching Proto this way, and instead we applied a big "sed" to all boost include files from libraries proto, phoenix, spirit and xpressive to inject in every of this header a pragma to silence the warning (because we still want to have this warning for other libraries, and for our own code). We were hoping that one day something would come with a solution, either by changing boost, the compilers, or even the C++ standard (because I think the compiler is only implementing what the standard says, it is indeed deprecated).

Reply:
> On our side, while this patch worked for us while we were using C++17 (maybe it was wrong, but in the use cases we had it did not show issue), however when we moved to C++20 indeed the fact that the "aggregate constructor" wasn't allowed anymore, it indeed broke the rest of the boost libraries using it.

I though that [P1008](http://wg21.link/p1008) will be applied retroactively, but in fact there is no even a warning in C++11/14/17 mode.

> We internally decided to stop patching Proto this way, and instead we applied a big "sed" to all boost include files from libraries proto, phoenix, spirit and xpressive to inject in every of this header a pragma to silence the warning (because we still want to have this warning for other libraries, and for our own code).

IIUC it should not have helped as the warning is issued at the use site and you have to disable it there, it *your* code. And this is the most cruel part about this GCC diagnostic.

> because I think the compiler is only implementing what the standard says, it is indeed deprecated

Yes, the warning itself is about what the standard says, but currently it is unusable because:
1. the deprecation was issued in C++11 where it was possible to fix it without making a type non-aggregate,
2. you cannot turn the warning on only for you code,
3. there are false-positives https://gcc.gnu.org/bugzilla/show_bug.cgi?id=92145

Reply:
> IIUC it should not have helped as the warning is issued at the use site and you have to disable it there, it your code. And this is the most cruel part about this GCC diagnostic.

Actually it did, because in all the use case I was shown, the call sites were inside the cited boost libraries. We never use ourselves boost proto, we use it only indirectly through other boost libraries.